python/lsst/ts/salobj/async_s3_bucket.py

```diff
@@ -0,0 +1,130 @@
+# This file is part of ts_salobj.
+#
+# Developed for the LSST Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["AsyncS3Bucket"]
+
+import asyncio
+import io
+
+import boto3
+import botocore
+
+
+class AsyncS3Bucket:
+    """Asynchronous interface to Amazon Web Services s3 buckets.
+
+    Parameters
+    ----------
+    name : `str`
+        Name of bucket.
+    """
+    def __init__(self, name):
+        s3 = boto3.resource("s3")
+        self.name = name
+        self.bucket = s3.Bucket(name)
+
+    async def upload(self, fileobj, key, callback=None):
+        """Upload a file-like object to the bucket.
+
+        Parameters
+        ----------
+        fileobj : file-like object
+            File-like object that can be read as *binary* data.
+        key : `str`
+            Name to use for the file in the bucket.
+        callback : callable (optional)
+            Function to call with updates while writing. The function receives
+            one argument: the number of bytes written. If the transfer is
+            successful then it will always be called at least once,
+            and the sum of the number of bytes for all calls will equal
+            the size of the file.
+
+        Notes
+        -----
+        To create a file-like object from an astropy.io.fits object:
+            fileobj = io.BytesIO()
```

Comment:
Did you forget to put a FITS object in here? I forget, would it be `io.BytesIO(hdulist)`?

Reply:
No. The full comment is:
```
        To create a file-like object from an astropy.io.fits object:
            fileobj = io.BytesIO()
            hdulist.writeto(fileobj)
```
But I'll amend the first sentence to:
```
To create a file-like object from an `astropy.io.fits` object named ``hdulist``
```

Reply:
Also:
* I added instructions for for converting a downloaded file-like object to an HDUList.
* Modified the download code to call `seek(0)` on the file-like object before returning it.
* Modified the round-trip test to use `read` instead of `getbuffer` in order to test that the downloaded file-like object has been rewound to the beginning.